cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -684,178 +621,185 @@ std::vector<std::vector<bool>> aggregate_reader_metadata::compute_data_page_mask
-    return {};  // An empty data page mask indicates all pages are required
-  }
 
-  // Compute page row counts, offsets, and column chunk page offsets for each column
-  std::vector<cudf::detail::host_vector<size_type>> page_row_counts;
-  std::vector<cudf::detail::host_vector<size_type>> page_row_offsets;
-  std::vector<cudf::detail::host_vector<size_type>> col_chunk_page_offsets;
-  page_row_counts.reserve(num_columns);
-  page_row_offsets.reserve(num_columns);
-  col_chunk_page_offsets.reserve(num_columns);
-
-  if (num_columns == 1) {
-    auto const schema_idx = column_schema_indices.front();
-    auto [counts, offsets, chunk_offsets] =
-      make_page_row_counts_and_offsets(per_file_metadata, row_group_indices, schema_idx, stream);
-    page_row_counts.emplace_back(std::move(counts));
-    page_row_offsets.emplace_back(std::move(offsets));
+  // Compute page row offsets and column chunk page offsets for each column
+  auto const num_columns = input_columns.size();
+  std::vector<size_type> page_row_offsets;
+  std::vector<size_type> col_page_offsets;
+  col_page_offsets.reserve(num_columns + 1);
+  col_page_offsets.push_back(0);
+
+  size_type max_page_size = 0;
+
+  if (num_columns <= 2) {
+    std::for_each(
+      column_schema_indices.begin(), column_schema_indices.end(), [&](auto const schema_idx) {
+        auto [col_page_row_offsets, col_max_page_size] =
+          compute_page_row_offsets(per_file_metadata, row_group_indices, schema_idx);
+        page_row_offsets.insert(
+          page_row_offsets.end(), col_page_row_offsets.begin(), col_page_row_offsets.end());
+        max_page_size = std::max<size_type>(max_page_size, col_max_page_size);
+        col_page_offsets.emplace_back(page_row_offsets.size());
+      });
   } else {
-    std::vector<std::future<std::tuple<cudf::detail::host_vector<size_type>,
-                                       cudf::detail::host_vector<size_type>,
-                                       cudf::detail::host_vector<size_type>>>>
-      page_row_counts_and_offsets_tasks;
-    page_row_counts_and_offsets_tasks.reserve(num_columns);
-
-    auto streams = cudf::detail::fork_streams(stream, num_columns);
-
-    std::for_each(thrust::counting_iterator<size_t>(0),
-                  thrust::counting_iterator(num_columns),
-                  [&](auto const col_idx) {
-                    page_row_counts_and_offsets_tasks.emplace_back(
-                      cudf::detail::host_worker_pool().submit_task([&, col_idx = col_idx] {
-                        return make_page_row_counts_and_offsets(per_file_metadata,
-                                                                row_group_indices,
-                                                                column_schema_indices[col_idx],
-                                                                streams[col_idx]);
-                      }));
-                  });
+    auto constexpr max_data_page_mask_threads = 2;
```

Comment:
Empirically use a max of 2 C++ threads instead of the host_worker_pool. 4 threads also work but was seeing performance regression beyond that